### PR TITLE
Add reconcile protection for roles

### PR DESCRIPTION
**Pull request by liggitt**
Allows customizing default roles and protecting them from reconciliation. This will let us safely customize admin/edit/view roles in environments like online

Proposed annotation name is `openshift.io/reconcile-protect` to align with proposed `kubernetes.io/write-protect` annotation upstream

### CODE REVIEW

**Comment by liggitt** [comment]
@deads2k

**Comment by liggitt** [comment]
[test]

File: test/fixtures/basic-user.json [hunk 1]
```diff
@@ -41,15 +41,6 @@
         },
         {
             "verbs": [
-                "list"
```
Comment (deads2k) [inline test/fixtures/basic-user.json]: nothing else used this file?

File: test/cmd/admin.sh [hunk 2]
```diff
@@ -177,6 +177,17 @@ os::cmd::expect_success 'oadm policy reconcile-cluster-roles clusterrole/cluster
 os::cmd::expect_success 'oc get clusterrole/cluster-status'
 
 os::cmd::expect_success 'oc replace --force -f ./test/fixtures/basic-user.json'
+# reconciliation protection
+# 1. mark the role as protected, and ensure the role is skipped by reconciliation
+os::cmd::expect_success 'oc annotate clusterrole/basic-user openshift.io/reconcile-protect=true'
+os::cmd::expect_success_and_text     'oadm policy reconcile-cluster-roles basic-user --additive-only=false' 'skipped: clusterrole/basic-user'
+# 2. unmark the role as protected, and ensure reconcile expects to remove extra permissions, and put back removed permissions
+os::cmd::expect_success 'oc annotate clusterrole/basic-user openshift.io/reconcile-protect=false --overwrite'
+os::cmd::expect_success_and_text     'oc get clusterrole/basic-user -o json'                                        'groups'
```
Comment (deads2k) [inline test/cmd/admin.sh]: json path template please?  I can't figure out what this is checking.

File: test/cmd/admin.sh [hunk 3]
```diff
@@ -177,6 +177,17 @@ os::cmd::expect_success 'oadm policy reconcile-cluster-roles clusterrole/cluster
...
+os::cmd::expect_success_and_not_text 'oadm policy reconcile-cluster-roles basic-user -o json --additive-only=false' 'groups'
+os::cmd::expect_success_and_not_text 'oc get clusterrole/basic-user -o json'                                        'projectrequests'
+os::cmd::expect_success_and_text     'oadm policy reconcile-cluster-roles basic-user -o json --additive-only=false' 'projectrequests'
```
Comment (deads2k) [inline test/cmd/admin.sh]: can you end by actually restoring it?

Comment (deads2k) [inline test/cmd/admin.sh]: > can you end by actually restoring it?

Oh, or does that happen later on?

**Comment by deads2k** [comment]
lgtm

File: test/fixtures/basic-user.json (hunk 1 above)
Comment (liggitt) [inline test/fixtures/basic-user.json]: split into a separate file for this test

File: test/cmd/admin.sh (hunk 2 above)
Comment (liggitt) [inline test/cmd/admin.sh]: done

File: test/cmd/admin.sh (hunk 3 above)
Comment (liggitt) [inline test/cmd/admin.sh]: happens later, added a restore at the end of each test stanza

**Comment by liggitt** [comment]
[merge]

**Pull request merged**